nixos/modules/services/security/fwknopd.nix

```diff
@@ -0,0 +1,287 @@
+              default = null;
+              description = "How many seconds to keep the access open.";
+            };
+            accessExpire = mkOption {
+              type = nullOr str;
+              default = null;
+              description = "Ignore knocks after MM/DD/YYYY.";
+            };
+            accessExpireEpoch = mkOption {
+              type = nullOr int;
+              default = null;
+              description = "Ignore knocks after Unix epoch time.";
+            };
+            cmdCycleOpen = mkOption {
+              type = nullOr str;
+              default = null;
+              description = "Command to execute on a successful knock.";
+            };
+            cmdCycleClose = mkOption {
+              type = nullOr str;
+              default = null;
+              description = "Command to execute on a successful knock, after cmdCycleTimer expires.";
+            };
+            cmdCycleTimer = mkOption {
+              type = nullOr int;
+              default = null;
+              description = "How many seconds to wait, before invoking cmdCycleClose.";
+            };
+            requireUsername = mkOption {
+              type = nullOr str;
+              default = null;
+              description = "Require a user name to be specified in the knock packet.";
+            };
+            requireSourceAddress = mkOption {
+              type = nullOr bool;
+              default = true;
+              description = "Require the real source address to be included in the packet (recommended).";
+            };
+          };
+        });
+
+        description = ''
+          Access grants.
+        '';
+        default = {};
+        example = {
+          "IP,..,IP/NET,..,NET/ANY" = {
+            fwAccessTimeout = 3600;
+            cmdCycleOpen = "echo 'Knock, knock!' | systemd-cat";
+          };
+        };
+      };
+    };
+  };
+
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    warnings = mkIf (cfg.test == true) [
+      "fwknopd operating in test mode!"
+    ];
+
+    users.users.fwknopd = {
+      home = "/var/lib/fwknopd";
+      group = "fwknopd";
+      isSystemUser = true;
+    };
+
+    users.groups.fwknopd = {};
+
+    systemd.targets.fwknop = {
+          wantedBy = [ "multi-user.target" ];
+          wants = [ "fwknopd.service" ];
+        };
+    systemd.services.fwknopd = {
+      description = "Single Packet Authorization server: fwknopd";
+      documentation = [ "man:fwknopd(8) http://www.cipherdyne.org/fwknop/" ];
+
+      enable = true;
+      after = [ "network-online.target" ];
+      wants = [ "network-online.target" ];
+
+      serviceConfig = {
+        User = "fwknopd";
+        Group = "fwknopd";
+        UMask = 0077;
+        StateDirectory = "fwknopd";
+        StateDirectoryMode = "700";
+        AmbientCapabilities = "CAP_NET_RAW CAP_NET_ADMIN";
+        Restart = "on-failure";
+        ExecStart = pkgs.writeShellScript "fwknopd-start.sh"
```

Comment:
Any reason this can't just be written as `ExecStart = toString [ "${fwknop}/bin/fwknopd" ... ];`

Reply:
@jmbaur, so you're essentially suggesting to inline the startup command line?

Reply:
Pretty much, it doesn't seem like there's anything happening here that would warrant the need for a shell in the ExecStart line.

Reply:
How about this?

```nix
ExecStart = "${pkgs.fwknop}/bin/fwknopd"
  + (lib.cli.toGNUCommandLineShell { } {
    inherit (cfg) interface test verbose;
    sudo-exe = "/run/wrappers/bin/sudo";
    foreground = true;
    run-dir = "/var/lib/fwknopd";
    config-file = configFile;
    access-file = configFile;
  })
```

Reply:
@h7x4, wow, that's really cool, thanks!

Reply:
Only needed an extra space on the first line.